WIP: add events to production calculator

3geek14 opened the pull request:
This PR seeks to add events to the production calculator. In Eevee Week, Raikou Research, and Entei Research, we saw buffs to Pokemons' stats, like skill rates. These impact Pokemons' productions, so SleepAPI's production calculator ideally would support this.

We don't know what sort of events Select Button will run in the future, or how they might influence things. This PR handles event buffs with a function passed to the production calculator that modifies a Pokemon. When there's no active event, this function is simply the identity function: `(mon: Pokemon) => mon`. For other events, this function is a little more complicated. The ingredient drop boost in Raikou and Entei was complicated enough to warrant a helper function, which is in `event-util.ts`.

Another approach would be to add more inputs to the production calcualtor to handle events. So far, we've only seen a small number of changes made: increased skill rate, increased ingredient drops, and restricting bonuses to a specific type. I don't like this approach, because I think it risks making it harder to maintain the codebase as new events have new mechanics. If a future event restricts some bonuses to one type and other bonuses to another type, then we'd have to rewrite anything that assumes all bonuses will either go to all Pokemon or to only Pokemon of a specific type.

**WIP Status:**

The first commit adds the `Event` data type, and it adds the events we've seen so far. Future commits will add the ability to pass an event into the production calculator, add an event parameter to the production calculator API route, and add an API route to get a list of all events.

3geek14 committed:
feat: add Event type

The Event type stores data that will pass into the
production calculator. This commit also adds
constants for the events we've had so far that
impacted the production calculator.

For Entei Research, the buff to main skill levels
is handled by the user manually increasing the
level.

SleepAPI commented on common/src/domain/event/event.ts:
```diff
@@ -0,0 +1,69 @@
+
+export interface Event {
+  name: string;
+  startDate: string;
```
What do you think about Date vs string here? 

Date would enforce a correct format and it would mean we don't have to parse them all to date when listing current/future events in the frontend. 

Using string means we can list the dates in frontend without formatting a Date. 

If we go with Date we should be able to create an event like 
`new Date('2024-01-31')`

3geek14 replied:
Thanks, I much prefer the `Date` type.

SleepAPI commented on common/src/domain/event/event.ts:
```diff
@@ -0,0 +1,69 @@
+  statModifier: (mon: Pokemon) => Pokemon;
+}
+
+export const ENTEI_RESEARCH: Event = {
```
I guess listing all the events in here will eventually become messy. Do you think we could take a similar approach to how we do pokemon? Perhaps one folder for each year, so right now we would have 2023 and 2024

3geek14 replied:
I'm not convinced this is necessary, as we've still got way way fewer events than Pokemon. I'd be inclined to punt this until it becomes more of an issue?

SleepAPI replied:
That's fine, this isn't a big issue to me

3geek14 commented on common/src/domain/event/event-util.ts:
```diff
@@ -0,0 +1,8 @@
+import { IngredientSet } from 'src/domain';
```
Done

3geek14 commented on common/src/domain/event/event.ts:
```diff
@@ -0,0 +1,69 @@
+import { AddOneIngredient } from 'src/domain/event/event-util';
+import { Pokemon } from 'src/domain/pokemon';
+
+export interface Event {
```
I've added some descriptions, but I'm not very attached to them. Happy to let other people change these later.

SleepAPI commented on common/src/domain/event/index.ts:
Since the domain/event folder is new it needs to be exported from src/domain as well. Currently this index file exports it out of domain/event, but the level above doesnt pick it up

webbywisp reviewed (changes requested):
One final comment from my side, regarding the export of domain/event

webbywisp reviewed (dismissed):
LGTM, just needs the rebase and merge

3geek14 committed:
WIP: refactor PokemonIngredientSet

In order to consume Event data in the production
calculator, the PokemonIngredientSet type is
replaced by the PokemonIngredientList type as an
input. This allows the production calculator to
modify a Pokemon due to an event before generating
its ingredient list.

This commit is not close to done, but I'm
uploading it so I can ask for help with the error
message.

3geek14 commented:
I want to push the event adjustment all the way into `setupAndRunProductionSimulation`. Doing that means changing the `PokemonIngredientSet` type to have "AAC" (for example) rather than a full `IngredientSet[]`. That would be a rather large refactor.

If I don't do that, it's easy to get `calculatePokemonProduction` in `production-service.ts` to use the `Event` it receives and pass the modified `Pokemon` into `setupAndRunProductionSimulation`. Similarly, `calculateOptimalProductionForSetCover` has a `ProductionStats` object and can make use of the `Event`.

However, `getAllOptimalIngredientFocusedPokemonProduce` in `contribution-calculator.ts` (used for the tier list) doesn't have a full `ProductionStats` object, so I can't do anything there. This is probably fine, but I think it'd be cool just as a curiosity to see what changes events make to the tier lists.

I think I'll try to do the refactoring. From a "don't repeat yourself" perspective, it feels wrong to deal with event modifications in multiple places. If tier lists don't want to deal with events, that's fine; they can simply not pass in any `Event`, and `setupAndRunProductionSimulation` can apply `NO_EVENT`.

3geek14 commented:
I started working on the refactor, and I ran into an error as soon as I tried to import `Event` into the backend. I've pushed the broken commit in case you want to be able to reproduce the issue. Here's the error message I got from `npm run build`:
```
There was a problem resolving type of 'Event'.
There was a problem resolving type of 'ProductionStats'.
Generate swagger error.
 GenerateMetadataError: Unknown type: FunctionType
At: ~/SleepAPI/common/dist/domain/event/event.d.ts:7:7.
This was caused by 'statModifier: (mon: Pokemon) => Pokemon;'
    at TypeResolver.resolve (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:480:19)
    at TypeResolver.propertyFromSignature (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1195:119)
    at ~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1160:39
    at Array.map (<anonymous>)
    at TypeResolver.getModelProperties (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1160:18)
    at TypeResolver.getModelReference (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1054:33)
    at calcReferenceType (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:920:50)
    at TypeResolver.getReferenceType (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:933:24)
    at TypeResolver.resolve (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:513:36)
    at TypeResolver.propertyFromSignature (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1195:119)
```

SleepAPI commented:
> I started working on the refactor, and I ran into an error as soon as I tried to import `Event` into the backend. I've pushed the broken commit in case you want to be able to reproduce the issue. Here's the error message I got from `npm run build`:
> 
> ```
> There was a problem resolving type of 'Event'.
> There was a problem resolving type of 'ProductionStats'.
> Generate swagger error.
>  GenerateMetadataError: Unknown type: FunctionType
> At: ~/SleepAPI/common/dist/domain/event/event.d.ts:7:7.
> This was caused by 'statModifier: (mon: Pokemon) => Pokemon;'
>     at TypeResolver.resolve (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:480:19)
>     at TypeResolver.propertyFromSignature (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1195:119)
>     at ~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1160:39
>     at Array.map (<anonymous>)
>     at TypeResolver.getModelProperties (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1160:18)
>     at TypeResolver.getModelReference (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1054:33)
>     at calcReferenceType (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:920:50)
>     at TypeResolver.getReferenceType (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:933:24)
>     at TypeResolver.resolve (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:513:36)
>     at TypeResolver.propertyFromSignature (~/SleepAPI/backend/node_modules/@tsoa/cli/dist/metadataGeneration/typeResolver.js:1195:119)
> ```

This is because ProductionStats interface is returned on the API for some routes. For example both the api/optimal/meal/{name} and your own api/optimal/ingredient/{name} return an attribute called filter of type `SetCoverProductionStats` which just extends ProductionStats. So Swagger is complaining that we can't return a function on the API, since Event type contains statModifier which is a typescript function and not compatible with a HTTP response

3geek14 commented:
Bluh...okay I understand the error now. I guess one option is to have the `ProductionStats` store a `string` that's the name of an `Event`? And then do a lookup. That would prevent power-users from writing their own custom event functions, which is probably desirable, to prevent us from running random user code that we don't trust.

SleepAPI commented:
> Bluh...okay I understand the error now. I guess one option is to have the `ProductionStats` store a `string` that's the name of an `Event`? And then do a lookup. That would prevent power-users from writing their own custom event functions, which is probably desirable, to prevent us from running random user code that we don't trust.

I think in general for the API incoming request we use ProductionRequest interface. So for incoming requests and lookup we can add an event identifier such as an id or name or something like ENTEI_2024 as id would be cool. 

I think the issue with ProductionStats is that today our APIs are too verbose, we probably shouldn't just return the entire ProductionStats interface out. I think we return some stuff like subskills, nature etc so that the old frontend could use it to list input settings on the results page. We probably should just add event to ProductionStats, but return more explicit type from the APIs. Might need a little refactoring though.

I believe the new frontend doesn't need the input settings returned at all and the team simulator doesn't. The new frontend sends the input settings to the backend and so it already has those beforehand ready.
